File: bindings/r/stridealign/R/keyboard.R
# R cannot place U+0000 in a CHARSXP. The keyboard facade therefore uses the
# representable ASCII code points 1..127; matrix index i still equals the ASCII
# code point i, and DEL remains the wildcard slot.
.stride_keyboard_ascii_symbols <- intToUtf8(1:127, multiple = TRUE)
.stride_keyboard_ascii_alphabet <- paste0(
  .stride_keyboard_ascii_symbols,
  collapse = ""
)
.stride_keyboard_ascii_wildcard <- intToUtf8(127)

.stride_keyboard_read_npy <- function(path) {
  connection <- file(path, open = "rb")
  on.exit(close(connection), add = TRUE)
  magic <- readBin(connection, what = "raw", n = 6L)
  if (!identical(magic, as.raw(c(0x93, charToRaw("NUMPY"))))) {
    stop("not a NumPy .npy file", call. = FALSE)
  }
  version <- as.integer(readBin(connection, what = "raw", n = 2L))
  header_size <- if (version[[1L]] == 1L) {
    readBin(
      connection, what = integer(), n = 1L, size = 2L,
      signed = FALSE, endian = "little"
    )
  } else if (version[[1L]] %in% c(2L, 3L)) {
    readBin(
      connection, what = integer(), n = 1L, size = 4L,
      signed = FALSE, endian = "little"
    )
  } else {
    stop("unsupported NumPy .npy version", call. = FALSE)
  }
  header <- rawToChar(readBin(connection, what = "raw", n = header_size))
  if (!grepl("['\"]descr['\"]\\s*:\\s*['\"]\\|i1['\"]", header, perl = TRUE)) {
    stop("keyboard .npy matrix must use signed int8 values", call. = FALSE)
  }
  if (grepl("['\"]fortran_order['\"]\\s*:\\s*True", header, perl = TRUE)) {
    stop("Fortran-order .npy matrices are not supported", call. = FALSE)
  }
  shape_match <- regexec(
    "['\"]shape['\"]\\s*:\\s*\\(([^)]*)\\)", header, perl = TRUE
  )
  captured <- regmatches(header, shape_match)[[1L]]
  if (length(captured) != 2L) {
    stop("could not read shape from keyboard .npy file", call. = FALSE)
  }
  dimensions <- as.integer(trimws(strsplit(captured[[2L]], ",")[[1L]]))
  dimensions <- dimensions[!is.na(dimensions)]
  if (length(dimensions) != 2L) {
    stop("keyboard .npy value must be a two-dimensional matrix", call. = FALSE)
  }
  payload <- readBin(connection, what = "raw", n = prod(dimensions))
  if (length(payload) != prod(dimensions)) {
    stop("truncated keyboard .npy payload", call. = FALSE)
  }
  values <- as.integer(payload)
  values[values >= 128L] <- values[values >= 128L] - 256L
  matrix(values, nrow = dimensions[[1L]], ncol = dimensions[[2L]], byrow = TRUE)
}

.stride_keyboard_grid_for_alphabet <- function(grid, alphabet) {
  size <- nchar(alphabet, type = "chars")
  # Bundled Python artifacts include the unrepresentable NUL row/column.
  if (identical(dim(grid), c(128L, 128L)) && size == 127L) {
    grid <- grid[2:128, 2:128, drop = FALSE]
  }
  if (!identical(dim(grid), c(size, size))) {
    stop(
      "keyboard matrix shape does not match the alphabet size",
      call. = FALSE
    )
  }
  grid
}

keyboard_from_npy <- function(
  path,
  name = tools::file_path_sans_ext(basename(path)),
  alphabet = keyboard$ASCII_ALPHABET,
  wildcard = keyboard$ASCII_WILDCARD,
  transpose = FALSE,
  gap_score = -1,
  gap_open = NULL,
  gap_extend = NULL
) {
  if (!is.character(path) || length(path) != 1L || is.na(path)) {
    stop("path must be one non-missing filename", call. = FALSE)
  }
  grid <- .stride_keyboard_grid_for_alphabet(
    .stride_keyboard_read_npy(path), alphabet
  )
  if (isTRUE(transpose)) grid <- t(grid)
  SubstitutionMatrix(
    name = name,
    alphabet = alphabet,
    matrix = grid,
    gap_score = gap_score,
    wildcard = wildcard,
    gap_open = gap_open,
    gap_extend = gap_extend
  )
}

.stride_keyboard_counts_grid <- function(counts, alphabet) {
  symbols <- strsplit(alphabet, "", fixed = TRUE)[[1L]]
  size <- length(symbols)
  if (is.matrix(counts)) {
    if (!is.numeric(counts) || !identical(dim(counts), c(size, size))) {
      stop("counts matrix shape does not match the alphabet", call. = FALSE)
    }
    if (anyNA(counts) || any(!is.finite(counts)) || any(counts < 0)) {
      stop("counts must be finite non-negative numbers", call. = FALSE)
    }
    return(matrix(as.numeric(counts), size, size))
  }
  if (!is.data.frame(counts) ||
      !all(c("typed", "intended", "count") %in% names(counts))) {
    stop(
      "counts must be a numeric matrix or a data frame with typed, intended, and count columns",
      call. = FALSE
    )
  }
  if (!is.character(counts$typed) || !is.character(counts$intended) ||
      !is.numeric(counts$count) || anyNA(counts$count) ||
      any(!is.finite(counts$count)) || any(counts$count < 0)) {
    stop("confusion-count rows are invalid", call. = FALSE)
  }
  grid <- matrix(0, size, size)
  typed <- match(counts$typed, symbols)
  intended <- match(counts$intended, symbols)
  keep <- !is.na(typed) & !is.na(intended)
  if (any(keep)) {
    for (index in which(keep)) {
      grid[typed[[index]], intended[[index]]] <- counts$count[[index]]
    }
  }
  grid
}

keyboard_from_confusion_counts <- function(
  counts,
  alphabet = keyboard$ASCII_ALPHABET,
  name = "KEYBOARD",
  scale = 2,
  match_margin = 4,
  floor = NULL,
  wildcard = substr(alphabet, nchar(alphabet), nchar(alphabet)),
  gap_score = -1,
  gap_open = NULL,
  gap_extend = NULL
) {
  grid_counts <- .stride_keyboard_counts_grid(counts, alphabet)
  size <- nrow(grid_counts)
  total <- sum(grid_counts)
  if (total == 0) {
    grid <- matrix(-1L, size, size)
    diag(grid) <- 1L
  } else {
    row_totals <- rowSums(grid_counts)
    column_totals <- colSums(grid_counts)
    denominator <- outer(row_totals, column_totals)
    scores <- scale * log2((grid_counts * total) / denominator)
    finite <- is.finite(scores)
    resolved_floor <- if (is.null(floor)) {
      if (any(finite)) base::floor(min(scores[finite])) else -1
    } else {
      floor
    }
    scores[!finite] <- resolved_floor
    grid <- matrix(
      pmax(-128, pmin(127, round(scores))),
      nrow = size,
      ncol = size
    )
    diag(grid) <- -128
    best_substitution <- max(grid)
    diag(grid) <- min(127, max(best_substitution, 0) + match_margin)
    storage.mode(grid) <- "integer"
  }
  SubstitutionMatrix(
    name = name,
    alphabet = alphabet,
    matrix = grid,
    gap_score = gap_score,
    wildcard = wildcard,
    gap_open = gap_open,
    gap_extend = gap_extend
  )
}

keyboard_available <- function() {
  directory <- system.file("keyboard_data", package = "stridealign")
  if (!nzchar(directory) || !dir.exists(directory)) return(character())
  sort(tools::file_path_sans_ext(basename(list.files(
    directory, pattern = "[.]npy$", full.names = FALSE
  ))))
}

keyboard <- new.env(parent = emptyenv())
keyboard$ASCII_ALPHABET <- .stride_keyboard_ascii_alphabet
keyboard$ASCII_CODEPOINTS <- 1:127
keyboard$ASCII_WILDCARD <- .stride_keyboard_ascii_wildcard
keyboard$from_confusion_counts <- keyboard_from_confusion_counts
keyboard$from_npy <- keyboard_from_npy
keyboard$available <- keyboard_available

.stride_keyboard_load_bundled <- function() {
  directory <- system.file("keyboard_data", package = "stridealign")
  for (name in keyboard_available()) {
    assign(
      name,
      keyboard_from_npy(
        file.path(directory, paste0(name, ".npy")),
        name = paste0("keyboard:", name)
      ),
      envir = keyboard
    )
  }
  invisible(NULL)
}

File: bindings/go/keyboard.go
package stridealign

import (
	"encoding/binary"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"unicode/utf8"
)

// KeyboardASCIIAlphabet contains all 128 ASCII code points, including NUL.
var KeyboardASCIIAlphabet = func() string {
	symbols := make([]rune, 128)
	for index := range symbols {
		symbols[index] = rune(index)
	}
	return string(symbols)
}()

const KeyboardASCIIWildcard rune = 127

// KeyboardOptions controls conversion of keyboard confusion counts.
type KeyboardOptions struct {
	Name           string
	Alphabet       string
	Wildcard       rune
	Scale          float64
	MatchMargin    float64
	Floor          *float64
	GapScore       int64
	GapOpenScore   int64
	GapExtendScore int64
	AffineGaps     bool
}

func DefaultKeyboardOptions() KeyboardOptions {
	return KeyboardOptions{
		Name: "KEYBOARD", Alphabet: KeyboardASCIIAlphabet,
		Wildcard: KeyboardASCIIWildcard, Scale: 2, MatchMargin: 4,
		GapScore: -1,
	}
}

// ConfusionCount records how often Typed appeared when Intended was expected.
type ConfusionCount struct {
	Typed    rune
	Intended rune
	Count    float64
}

func keyboardOptions(options *KeyboardOptions) KeyboardOptions {
	resolved := DefaultKeyboardOptions()
	if options != nil {
		resolved = *options
	}
	return resolved
}

func keyboardFromGrid(counts [][]float64, options *KeyboardOptions) (*SubstitutionMatrix, error) {
	resolved := keyboardOptions(options)
	symbols := []rune(resolved.Alphabet)
	size := len(symbols)
	if len(counts) != size {
		return nil, fmt.Errorf("stridealign: keyboard count rows do not match alphabet")
	}
	total := 0.0
	rowTotals := make([]float64, size)
	columnTotals := make([]float64, size)
	for row, values := range counts {
		if len(values) != size {
			return nil, fmt.Errorf("stridealign: keyboard count columns do not match alphabet")
		}
		for column, value := range values {
			if math.IsNaN(value) || math.IsInf(value, 0) || value < 0 {
				return nil, fmt.Errorf("stridealign: keyboard counts must be finite and non-negative")
			}
			total += value
			rowTotals[row] += value
			columnTotals[column] += value
		}
	}
	scores := make([][]int, size)
	for row := range scores {
		scores[row] = make([]int, size)
	}
	if total == 0 {
		for row := range scores {
			for column := range scores[row] {
				scores[row][column] = -1
			}
			scores[row][row] = 1
		}
	} else {
		computed := make([][]float64, size)
		floor := math.Inf(1)
		for row := range computed {
			computed[row] = make([]float64, size)
			for column := range computed[row] {
				denominator := rowTotals[row] * columnTotals[column]
				value := math.Inf(-1)
				if counts[row][column] > 0 && denominator > 0 {
					value = resolved.Scale * math.Log2(counts[row][column]*total/denominator)
					if value < floor {
						floor = value
					}
				}
				computed[row][column] = value
			}
		}
		if resolved.Floor != nil {
			floor = *resolved.Floor
		} else if math.IsInf(floor, 1) {
			floor = -1
		} else {
			floor = math.Floor(floor)
		}
		bestSubstitution := -128
		for row := range computed {
			for column, value := range computed[row] {
				if math.IsInf(value, 0) || math.IsNaN(value) {
					value = floor
				}
				score := int(math.Round(value))
				if score < -128 {
					score = -128
				} else if score > 127 {
					score = 127
				}
				scores[row][column] = score
				if row != column && score > bestSubstitution {
					bestSubstitution = score
				}
			}
		}
		match := int(math.Round(math.Max(float64(bestSubstitution), 0) + resolved.MatchMargin))
		if match > 127 {
			match = 127
		}
		for index := range scores {
			scores[index][index] = match
		}
	}
	matrixOptions := MatrixOptions{
		Wildcard: resolved.Wildcard, GapScore: resolved.GapScore,
		GapOpenScore:   resolved.GapOpenScore,
		GapExtendScore: resolved.GapExtendScore, AffineGaps: resolved.AffineGaps,
	}
	return NewSubstitutionMatrix(resolved.Name, resolved.Alphabet, scores, &matrixOptions)
}

// KeyboardFromConfusionGrid builds a matrix from a dense typed-by-intended grid.
func KeyboardFromConfusionGrid(counts [][]float64, options *KeyboardOptions) (*SubstitutionMatrix, error) {
	return keyboardFromGrid(counts, options)
}

// KeyboardFromConfusionCounts builds a matrix from sparse observations.
func KeyboardFromConfusionCounts(counts []ConfusionCount, options *KeyboardOptions) (*SubstitutionMatrix, error) {
	resolved := keyboardOptions(options)
	symbols := []rune(resolved.Alphabet)
	indices := make(map[rune]int, len(symbols))
	for index, symbol := range symbols {
		indices[symbol] = index
	}
	grid := make([][]float64, len(symbols))
	for index := range grid {
		grid[index] = make([]float64, len(symbols))
	}
	for _, count := range counts {
		if math.IsNaN(count.Count) || math.IsInf(count.Count, 0) || count.Count < 0 {
			return nil, fmt.Errorf("stridealign: keyboard counts must be finite and non-negative")
		}
		typed, typedOK := indices[count.Typed]
		intended, intendedOK := indices[count.Intended]
		if typedOK && intendedOK {
			grid[typed][intended] += count.Count
		}
	}
	return keyboardFromGrid(grid, &resolved)
}

var (
	npyDescrPattern   = regexp.MustCompile(`['"]descr['"]\s*:\s*['"]\|i1['"]`)
	npyFortranPattern = regexp.MustCompile(`['"]fortran_order['"]\s*:\s*True`)
	npyShapePattern   = regexp.MustCompile(`['"]shape['"]\s*:\s*\(([^)]*)\)`)
)

func readInt8NPY(path string) ([][]int, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	if len(data) < 10 || string(data[:6]) != "\x93NUMPY" {
		return nil, fmt.Errorf("stridealign: not a NumPy .npy file")
	}
	major := data[6]
	headerStart := 10
	headerSize := 0
	if major == 1 {
		headerSize = int(binary.LittleEndian.Uint16(data[8:10]))
	} else if major == 2 || major == 3 {
		if len(data) < 12 {
			return nil, fmt.Errorf("stridealign: truncated NumPy header")
		}
		headerStart = 12
		headerSize = int(binary.LittleEndian.Uint32(data[8:12]))
	} else {
		return nil, fmt.Errorf("stridealign: unsupported NumPy version %d", major)
	}
	if headerSize < 0 || headerStart+headerSize > len(data) {
		return nil, fmt.Errorf("stridealign: truncated NumPy header")
	}
	header := string(data[headerStart : headerStart+headerSize])
	if !npyDescrPattern.MatchString(header) {
		return nil, fmt.Errorf("stridealign: keyboard .npy matrix must use signed int8 values")
	}
	if npyFortranPattern.MatchString(header) {
		return nil, fmt.Errorf("stridealign: Fortran-order .npy matrices are unsupported")
	}
	match := npyShapePattern.FindStringSubmatch(header)
	if len(match) != 2 {
		return nil, fmt.Errorf("stridealign: cannot read NumPy shape")
	}
	var dimensions []int
	for _, field := range strings.Split(match[1], ",") {
		field = strings.TrimSpace(field)
		if field == "" {
			continue
		}
		value, parseErr := strconv.Atoi(field)
		if parseErr != nil || value < 0 {
			return nil, fmt.Errorf("stridealign: invalid NumPy shape")
		}
		dimensions = append(dimensions, value)
	}
	if len(dimensions) != 2 || dimensions[0] != dimensions[1] {
		return nil, fmt.Errorf("stridealign: keyboard .npy value must be square and two-dimensional")
	}
	payload := data[headerStart+headerSize:]
	if dimensions[0] != 0 && dimensions[1] > int(^uint(0)>>1)/dimensions[0] {
		return nil, fmt.Errorf("stridealign: NumPy shape is too large")
	}
	count := dimensions[0] * dimensions[1]
	if len(payload) != count {
		return nil, fmt.Errorf("stridealign: NumPy payload size does not match shape")
	}
	grid := make([][]int, dimensions[0])
	for row := range grid {
		grid[row] = make([]int, dimensions[1])
		for column := range grid[row] {
			grid[row][column] = int(int8(payload[row*dimensions[1]+column]))
		}
	}
	return grid, nil
}

// KeyboardFromNPY loads the signed-int8 C-order format used by Python's keyboard matrices.
func KeyboardFromNPY(path string, transpose bool, options *KeyboardOptions) (*SubstitutionMatrix, error) {
	grid, err := readInt8NPY(path)
	if err != nil {
		return nil, err
	}
	resolved := keyboardOptions(options)
	if options == nil {
		resolved.Name = strings.TrimSuffix(filepath.Base(path), filepath.Ext(path))
	}
	size := utf8.RuneCountInString(resolved.Alphabet)
	if len(grid) != size {
		return nil, fmt.Errorf("stridealign: keyboard matrix shape does not match alphabet")
	}
	if transpose {
		for row := 0; row < size; row++ {
			for column := row + 1; column < size; column++ {
				grid[row][column], grid[column][row] = grid[column][row], grid[row][column]
			}
		}
	}
	matrixOptions := MatrixOptions{
		Wildcard: resolved.Wildcard, GapScore: resolved.GapScore,
		GapOpenScore:   resolved.GapOpenScore,
		GapExtendScore: resolved.GapExtendScore, AffineGaps: resolved.AffineGaps,
	}
	return NewSubstitutionMatrix(resolved.Name, resolved.Alphabet, grid, &matrixOptions)
}

// KeyboardAvailable lists embedded keyboard matrices.
func KeyboardAvailable() ([]string, error) {
	all, err := AvailableMatrices()
	if err != nil {
		return nil, err
	}
	var names []string
	for _, name := range all {
		if strings.HasPrefix(name, "keyboard:") {
			names = append(names, strings.TrimPrefix(name, "keyboard:"))
		}
	}
	sort.Strings(names)
	return names, nil
}

// KeyboardMatrix loads one embedded keyboard matrix by short name.
func KeyboardMatrix(name string) (*SubstitutionMatrix, error) {
	return BuiltinMatrix("keyboard:" + strings.TrimPrefix(name, "keyboard:"))
}
